orttraining/orttraining/core/session/training_session.cc

```diff
@@ -959,9 +969,12 @@ common::Status TrainingSession::GetOptimizerState(std::unordered_map<std::string
   // Change key from sharded_name to weight_name using partition_info
   for (const auto& weight : weight_partition_info_) {
     const auto& it = opt_state_tensors.find(weight.second.view_name);
-    ORT_ENFORCE(it != opt_state_tensors.end(), "Cannot find weight: " + weight.second.view_name + " in weight_partition_info_");
-    opt_state_tensors[weight.first] = it->second;
-    opt_state_tensors.erase(it);
+    if (it == opt_state_tensors.end()) {
+      ORT_RETURN_IF_NOT(allow_missing, "Failed to get optimizer params for partition: " + weight.second.view_name);
```

Comment:
Is this just for zero partition?

Reply:
that is correct, as for Megatron, there must be a state present for all partitioed/non-partitioned weights/optimizer states. Missing initializers are allowed only when Zero is used.